REGRESSION(264618@main): Infinite mutual recursion in SVGRenderSupport::layoutDifferentRootIfNeeded

rniwa opened the pull request:
#### 0d62c4b9488c26e3e1c97771e9cb06c397fe5fad
<pre>
REGRESSION(264618@main): Infinite mutual recursion in SVGRenderSupport::layoutDifferentRootIfNeeded
<a href="https://bugs.webkit.org/show_bug.cgi?id=258171">https://bugs.webkit.org/show_bug.cgi?id=258171</a>

Reviewed by Darin Adler.

Prevent infinite recursion in SVGRenderSupport::layoutDifferentRootIfNeeded by checking
whether the target root is already in layout or not. Also removed the now unused svgRoot argument
from layoutDifferentRootIfNeeded and renamed the function to layoutReferencedRootIfNeeded.

* LayoutTests/svg/markers/marker-clip-path-mutual-reference-crash-expected.txt: Added.
* LayoutTests/svg/markers/marker-clip-path-mutual-reference-crash.html: Added.
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
(WebCore::SVGContainerLayout::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::layoutReferencedRootIfNeeded): Renamed from layoutDifferentRootIfNeeded.
* Source/WebCore/rendering/svg/SVGResources.h:

Canonical link: <a href="https://commits.webkit.org/265249@main">https://commits.webkit.org/265249@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d25081db639cc3ca8f04d9a80992ee685397a806

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10553 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11962 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9919 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10328 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12551 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10503 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12876 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10474 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11209 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8692 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12348 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8516 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9346 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16618 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9614 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9497 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12746 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9937 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8067 "4 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9096 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9241 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2491 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13347 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -313,23 +313,20 @@ bool SVGResources::buildCachedResources(const RenderElement& renderer, const Ren
+        if (!container)
+            return;
+        auto* root = SVGRenderSupport::findTreeRootObject(*container);
+        if (svgRoot == root || root->isInLayout())
```
Do we even need the `svgRoot == root` check any more?

rniwa replied:
No, we don't. In fact, we can just get rid of svgRoot argument altogether now.

shallawa commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -313,23 +313,20 @@ bool SVGResources::buildCachedResources(const RenderElement& renderer, const Ren
+            return;
+        auto* root = SVGRenderSupport::findTreeRootObject(*container);
+        if (svgRoot == root || root->isInLayout())
+            return;
```
Why do we need to add this check? Does not SVGResourcesCycleSolver::resolveCycles() break the cycle between the SVG resources?

rniwa replied:
SVGResourcesCycleSolver::resolveCycles DOES break the cycle. What we need to prevent here is the (mutual) recursion of SVGRenderSupport::layoutDifferentRootIfNeeded -> SVGResources::layoutDifferentRootIfNeeded -> RenderElement::layoutIfNeeded -> 
RenderSVGResourceContainer::layout -> LegacyRenderSVGHiddenContainer::layout -> SVGRenderSupport::layoutChildren -> SVGRenderSupport::layoutDifferentRootIfNeeded. This recursion exists independent of SVG resource cycles (when there are more than two SVG root elements whose descendants mutually reference each other) so we need to manually prevent infinite recursion here.

rniwa committed:
REGRESSION(264618@main): Infinite mutual recursion in SVGRenderSupport::layoutDifferentRootIfNeeded
https://bugs.webkit.org/show_bug.cgi?id=258171

Reviewed by Darin Adler.

Prevent infinite recursion in SVGRenderSupport::layoutDifferentRootIfNeeded by checking
whether the target root is already in layout or not. Also removed the now unused svgRoot argument
from layoutDifferentRootIfNeeded and renamed the function to layoutReferencedRootIfNeeded.

* LayoutTests/svg/markers/marker-clip-path-mutual-reference-crash-expected.txt: Added.
* LayoutTests/svg/markers/marker-clip-path-mutual-reference-crash.html: Added.
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
(WebCore::SVGContainerLayout::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::layoutDifferentRootIfNeeded):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::layoutReferencedRootIfNeeded): Renamed from layoutDifferentRootIfNeeded.
* Source/WebCore/rendering/svg/SVGResources.h:

Canonical link: https://commits.webkit.org/265249@main
